### PR TITLE
[TensorIR] Primitive "SetScope"

### PR DESCRIPTION
**Pull request by MasterJH5574**
This PR introduces "SetScope", a schedule primitive which helps update the storage scope of a buffer.

cc @Hzfengsy @junrushao1994

### CODE REVIEW

File: src/tir/schedule/primitive/block_annotate.cc [hunk 1]
```diff
@@ -16,6 +16,9 @@
  * specific language governing permissions and limitations
  * under the License.
  */
+#include <tvm/runtime/container/array.h>
```
Comment (Hzfengsy) [inline src/tir/schedule/primitive/block_annotate.cc:19]: Do we need to include this explictly?

File: src/tir/schedule/primitive/block_annotate.cc [hunk 2]
```diff
@@ -233,14 +238,139 @@ class StorageAlignInvalidAnnotationError : public ScheduleError {
...
+      Block new_block(n);
+      block_sref_reuse_->Set(GetRef<Block>(block), new_block);
+      return new_block;
+    }
+  }
+
+  /*! \brief The storage scope to be set. */
+  String storage_scope_;
+  /*! \brief A mapping which maps old buffer vars to new buffers, including the buffers defined in
+   *         MatchBufferRegion.*/
```
Comment (Hzfengsy) [inline src/tir/schedule/primitive/block_annotate.cc:360]:
```suggestion
  /*!
   * \brief A mapping which maps old buffer vars to new buffers, including the buffers defined in
   *         MatchBufferRegion.
   */
```

**Review by Hzfengsy — CHANGES_REQUESTED** [review]
Please add a testcase for invalid storage scope name.

File: src/tir/schedule/primitive/block_annotate.cc (hunk 2 above)
Comment (MasterJH5574) [reply src/tir/schedule/primitive/block_annotate.cc]: Updated. Thanks for pointing out!

File: src/tir/schedule/analysis.h [hunk 3]
```diff
@@ -393,6 +393,10 @@ std::vector<runtime::TypedPackedFunc<CommReducer(DataType)>> GetReducerGetters()
 bool FromIdentityCombiner(const PrimExpr& identity, const BufferStore& combiner,
                           CommReducer* result_reducer, PrimExpr* lhs, PrimExpr* rhs);
 
+/******** Storage Scope ********/
+
+void CheckStorageScope(const ScheduleState& self, String storage_scope);
```
Comment (junrushao1994) [inline src/tir/schedule/analysis.h:398]: Doc?

File: src/tir/schedule/primitive/block_annotate.cc (hunk 1 above)
Comment (junrushao1994) [reply src/tir/schedule/primitive/block_annotate.cc:19]: Actually not. See utils.h

Comment (MasterJH5574) [reply src/tir/schedule/primitive/block_annotate.cc:19]: Yeah you're right. But it's conventional to put `#include "../../ir/functor_common.h"` before `#include "../utils.h"`, in which case Array wasn't imported be fore `functor_common.h`.. I want to know that, is it acceptable to put `utils.h` before `functor_common.h`, or is there any other way 😥 ?

Comment (junrushao1994) [reply src/tir/schedule/primitive/block_annotate.cc:19]: You should add this include in functor_common.h

Comment (MasterJH5574) [reply src/tir/schedule/primitive/block_annotate.cc:19]: Ah that makes sense, thanks!

File: src/tir/schedule/analysis.h (hunk 3 above)
Comment (MasterJH5574) [reply src/tir/schedule/analysis.h:404]: Updated! Thanks for the reminder!

**Comment by junrushao1994** [comment]
For context, this primitive is only used in one of our rule Cross-Thread-Reduction. Welcome everyone to review! CC: @Lunderberg @csullivan

**Review by Hzfengsy — APPROVED** [review]

**Pull request merged**